[PyTorch] Parallelize gelu via tensoriterator

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58950 [PyTorch] Parallelize gelu via tensoriterator**
* #59334 [PyTorch Operator Benchmark] gelu benchmark
* #58949 [PyTorch] Expose interface to set grain size on cpu_kernel_vec

Use tensor iterator's API to set grain size in order to parallelize gelu op.

Benchmarks:
Server: core i9 8 cores 16 threads (mac laptop)

Size | &nbsp; | &nbsp; | &nbsp; | &nbsp; | (us) | &nbsp; | &nbsp;
-- | -- | -- | -- | -- | -- | -- | --
n | c | h | w | # of elements | No multithreading | Threads=4 | Speedup
1 | 3 | 32 | 32 | 3072 | 12.549 | 10.526 | 1.19219077
1 | 3 | 32 | 64 | 6144 | 15.26 | 17.027 | 0.89622364
1 | 3 | 32 | 128 | 12288 | 27.934 | 40.674 | 0.68677779
1 | 3 | 64 | 32 | 6144 | 14.803 | 22.408 | 0.66061228
1 | 3 | 64 | 64 | 12288 | 28.051 | 30.728 | 0.91288076
1 | 3 | 64 | 128 | 24576 | 76.549 | 40.075 | 1.91014348
1 | 3 | 128 | 32 | 12288 | 28.977 | 33.688 | 0.86015792
1 | 3 | 128 | 64 | 24576 | 52.805 | 37.098 | 1.4233921
1 | 3 | 128 | 128 | 49152 | 103.568 | 53.207 | 1.9465108
4 | 3 | 32 | 32 | 12288 | 28.859 | 30.304 | 0.95231653
4 | 3 | 32 | 64 | 24576 | 76.31 | 38.491 | 1.98254137
4 | 3 | 32 | 128 | 49152 | 102.799 | 59.081 | 1.73996716
4 | 3 | 64 | 32 | 24576 | 54.207 | 39.439 | 1.37445168
4 | 3 | 64 | 64 | 49152 | 103.329 | 55.11 | 1.87495917
4 | 3 | 64 | 128 | 98304 | 228.303 | 87.973 | 2.59514851
4 | 3 | 128 | 32 | 49152 | 105.258 | 54.244 | 1.94045424
4 | 3 | 128 | 64 | 98304 | 207.008 | 84.974 | 2.43613341
4 | 3 | 128 | 128 | 196608 | 405.651 | 153.414 | 2.64415894

Mobile: Samsung S8 US ARM A73






System info | Samsung S8 | US | 4 + 4 Arm73 | &nbsp; | p50(ms)
-- | -- | -- | -- | -- | --
n | c | # elements | No threading | Threads=4 | Speedup
1 | 2048 | 2048 | 0.061 | 0.076 | 0.802631579
1 | 4096 | 4096 | 0.093 | 0.097 | 0.958762887
1 | 8192 | 8192 | 0.158 | 0.149 | 1.060402685













Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

kimishpatel committed:
[PyTorch] Parallelize gelu via tensoriterator

Use tensor iterator's API to set grain size in order to parallelize gelu op.

Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PyTorch] Parallelize gelu via tensoriterator"


Use tensor iterator's API to set grain size in order to parallelize gelu op.

Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -270,6 +270,11 @@ void GeluKernelImpl(TensorIterator& it) {
     });
   } else {
+    auto grain_size = at::internal::GRAIN_SIZE;
+    constexpr int64_t GELU_MIN_ELEMENTS_FOR_MULTI_THREADING{128};
```
This needs to be adjusted depending on experiment.

kimishpatel committed:
Update on "[PyTorch] Parallelize gelu via tensoriterator"


Use tensor iterator's API to set grain size in order to parallelize gelu op.

Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PyTorch] Parallelize gelu via tensoriterator"


Use tensor iterator's API to set grain size in order to parallelize gelu op.

Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]

ezyang commented on test/test_nn.py:
```diff
@@ -8281,6 +8281,12 @@ def _gelu_ref(X):
                 _test_gelu(n, m, torch.float64, True)
                 _test_gelu(n, m, torch.float64, False)
 
+        # Tesst multi threaded
```
nit: Test

ezyang commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -270,6 +270,18 @@ void GeluKernelImpl(TensorIteratorBase& it) {
+    // Benchmarked on S8 US phone.
+    constexpr int64_t GELU_MIN_ELEMENTS_FOR_MULTI_THREADING{6144};
+#else
+    // Benchmarked on i9 8 core 16 thread machine.
```
A longer description of benchmark methodology would be appreciated when someone wants to update this later

ezyang commented on test/test_nn.py:
```diff
@@ -8281,6 +8281,12 @@ def _gelu_ref(X):
+        num_threads = torch.get_num_threads()
+        torch.set_num_threads(4)
+        _test_gelu(32, 32, torch.float32, False)
+        torch.set_num_threads(num_threads)
```
cc @mruberry

mruberry replied:
Thanks for the fyi. Creating a context manager for this would be nicer. Note that if the test fails the number of threads won't be reset properly here. Although this PR doesn't need to add a context manager, the reset to the original `num_threads` should happen in a `finally` block:

```
num_threads = torch.get_num_threads()
torch.set_num_threads(4)
try:
  _test_gelu(32, 32, torch.float32, False)
finally:
  torch.set_num_threads(num_threads)
```

kimishpatel replied:
Thats a good point. Thanks for the suggestion. Will fix.

ezyang reviewed (approved):
okey dokey

kimishpatel committed:
Update on "[PyTorch] Parallelize gelu via tensoriterator"


Use tensor iterator's API to set grain size in order to parallelize gelu op.

Benchmarks:
Server: core i9 8 cores 16 threads (mac laptop)

Size | &nbsp; | &nbsp; | &nbsp; | &nbsp; | (us) | &nbsp; | &nbsp;
-- | -- | -- | -- | -- | -- | -- | --
n | c | h | w | # of elements | No multithreading | Threads=4 | Speedup
1 | 3 | 32 | 32 | 3072 | 12.549 | 10.526 | 1.19219077
1 | 3 | 32 | 64 | 6144 | 15.26 | 17.027 | 0.89622364
1 | 3 | 32 | 128 | 12288 | 27.934 | 40.674 | 0.68677779
1 | 3 | 64 | 32 | 6144 | 14.803 | 22.408 | 0.66061228
1 | 3 | 64 | 64 | 12288 | 28.051 | 30.728 | 0.91288076
1 | 3 | 64 | 128 | 24576 | 76.549 | 40.075 | 1.91014348
1 | 3 | 128 | 32 | 12288 | 28.977 | 33.688 | 0.86015792
1 | 3 | 128 | 64 | 24576 | 52.805 | 37.098 | 1.4233921
1 | 3 | 128 | 128 | 49152 | 103.568 | 53.207 | 1.9465108
4 | 3 | 32 | 32 | 12288 | 28.859 | 30.304 | 0.95231653
4 | 3 | 32 | 64 | 24576 | 76.31 | 38.491 | 1.98254137
4 | 3 | 32 | 128 | 49152 | 102.799 | 59.081 | 1.73996716
4 | 3 | 64 | 32 | 24576 | 54.207 | 39.439 | 1.37445168
4 | 3 | 64 | 64 | 49152 | 103.329 | 55.11 | 1.87495917
4 | 3 | 64 | 128 | 98304 | 228.303 | 87.973 | 2.59514851
4 | 3 | 128 | 32 | 49152 | 105.258 | 54.244 | 1.94045424
4 | 3 | 128 | 64 | 98304 | 207.008 | 84.974 | 2.43613341
4 | 3 | 128 | 128 | 196608 | 405.651 | 153.414 | 2.64415894

Mobile: Samsung S8 US ARM A73






System info | Samsung S8 | US | 4 + 4 Arm73 | &nbsp; | p50(ms)
-- | -- | -- | -- | -- | --
n | c | # elements | No threading | Threads=4 | Speedup
1 | 2048 | 2048 | 0.061 | 0.076 | 0.802631579
1 | 4096 | 4096 | 0.093 | 0.097 | 0.958762887
1 | 8192 | 8192 | 0.158 | 0.149 | 1.060402685













Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]

kimishpatel committed:
Update on "[PyTorch] Parallelize gelu via tensoriterator"


Use tensor iterator's API to set grain size in order to parallelize gelu op.

Benchmarks:
Server: core i9 8 cores 16 threads (mac laptop)

Size | &nbsp; | &nbsp; | &nbsp; | &nbsp; | (us) | &nbsp; | &nbsp;
-- | -- | -- | -- | -- | -- | -- | --
n | c | h | w | # of elements | No multithreading | Threads=4 | Speedup
1 | 3 | 32 | 32 | 3072 | 12.549 | 10.526 | 1.19219077
1 | 3 | 32 | 64 | 6144 | 15.26 | 17.027 | 0.89622364
1 | 3 | 32 | 128 | 12288 | 27.934 | 40.674 | 0.68677779
1 | 3 | 64 | 32 | 6144 | 14.803 | 22.408 | 0.66061228
1 | 3 | 64 | 64 | 12288 | 28.051 | 30.728 | 0.91288076
1 | 3 | 64 | 128 | 24576 | 76.549 | 40.075 | 1.91014348
1 | 3 | 128 | 32 | 12288 | 28.977 | 33.688 | 0.86015792
1 | 3 | 128 | 64 | 24576 | 52.805 | 37.098 | 1.4233921
1 | 3 | 128 | 128 | 49152 | 103.568 | 53.207 | 1.9465108
4 | 3 | 32 | 32 | 12288 | 28.859 | 30.304 | 0.95231653
4 | 3 | 32 | 64 | 24576 | 76.31 | 38.491 | 1.98254137
4 | 3 | 32 | 128 | 49152 | 102.799 | 59.081 | 1.73996716
4 | 3 | 64 | 32 | 24576 | 54.207 | 39.439 | 1.37445168
4 | 3 | 64 | 64 | 49152 | 103.329 | 55.11 | 1.87495917
4 | 3 | 64 | 128 | 98304 | 228.303 | 87.973 | 2.59514851
4 | 3 | 128 | 32 | 49152 | 105.258 | 54.244 | 1.94045424
4 | 3 | 128 | 64 | 98304 | 207.008 | 84.974 | 2.43613341
4 | 3 | 128 | 128 | 196608 | 405.651 | 153.414 | 2.64415894

Mobile: Samsung S8 US ARM A73






System info | Samsung S8 | US | 4 + 4 Arm73 | &nbsp; | p50(ms)
-- | -- | -- | -- | -- | --
n | c | # elements | No threading | Threads=4 | Speedup
1 | 2048 | 2048 | 0.061 | 0.076 | 0.802631579
1 | 4096 | 4096 | 0.093 | 0.097 | 0.958762887
1 | 8192 | 8192 | 0.158 | 0.149 | 1.060402685













Differential Revision: [D28689819](https://our.internmc.facebook.com/intern/diff/D28689819/)

[ghstack-poisoned]
